File: scripts/build-benchmark-matrix.py
#!/usr/bin/env python3

import argparse
import json
import pathlib


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--task-list-file", required=True)
    parser.add_argument("--max-shards", type=int, default=3)
    return parser.parse_args()


def load_task_paths(task_list_file: pathlib.Path) -> list[str]:
    if not task_list_file.exists():
        return []
    return [line.strip() for line in task_list_file.read_text(encoding="utf-8").splitlines() if line.strip()]


def chunk_task_paths(task_paths: list[str], max_shards: int) -> list[list[str]]:
    if not task_paths:
        return []
    shard_count = max(1, min(max_shards, len(task_paths)))
    shards: list[list[str]] = [[] for _ in range(shard_count)]
    for index, task_path in enumerate(task_paths):
        shards[index % shard_count].append(task_path)
    return [shard for shard in shards if shard]


def build_matrix(task_paths: list[str], max_shards: int) -> list[dict[str, object]]:
    include: list[dict[str, object]] = []
    for shard_index, shard_paths in enumerate(chunk_task_paths(task_paths, max_shards), start=1):
        include.append(
            {
                "shard_index": shard_index,
                "task_files": "\n".join(shard_paths),
                "task_count": len(shard_paths),
            }
        )
    return include


def main() -> None:
    args = parse_args()
    task_paths = load_task_paths(pathlib.Path(args.task_list_file))
    print(json.dumps(build_matrix(task_paths, args.max_shards), ensure_ascii=False))


if __name__ == "__main__":
    main()

File: scripts/build-benchmark-matrix.mjs
#!/usr/bin/env node
// build-benchmark-matrix: shard a task list across N matrix entries.
// Node port of scripts/build-benchmark-matrix.py — no Python.
import { readFileSync, existsSync } from 'node:fs';
import { isMain } from './bench/lib.mjs';

export function loadTaskPaths(taskListFile) {
  if (!existsSync(taskListFile)) return [];
  const raw = readFileSync(taskListFile, 'utf-8');
  return raw.split('\n').map((l) => l.trim()).filter((l) => l.length > 0);
}

export function chunkTaskPaths(taskPaths, maxShards) {
  if (!taskPaths || taskPaths.length === 0) return [];
  const shardCount = Math.max(1, Math.min(maxShards, taskPaths.length));
  const shards = Array.from({ length: shardCount }, () => []);
  taskPaths.forEach((p, i) => shards[i % shardCount].push(p));
  return shards.filter((s) => s.length > 0);
}

export function buildMatrix(taskPaths, maxShards) {
  const chunks = chunkTaskPaths(taskPaths, maxShards);
  return chunks.map((shard, i) => ({
    shard_index: i + 1,
    task_files: shard.join('\n'),
    task_count: shard.length,
  }));
}

function parseArgs(argv) {
  const out = { taskListFile: null, maxShards: 3 };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--task-list-file') { out.taskListFile = argv[++i]; }
    else if (a === '--max-shards') { out.maxShards = Number.parseInt(argv[++i], 10); }
    else { process.stderr.write(`unknown argument: ${a}\n`); process.exit(2); }
  }
  if (!out.taskListFile) { process.stderr.write('Usage: build-benchmark-matrix.mjs --task-list-file FILE [--max-shards N]\n'); process.exit(2); }
  return out;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const paths = loadTaskPaths(args.taskListFile);
  process.stdout.write(JSON.stringify(buildMatrix(paths, args.maxShards), null, 0));
}

if (isMain(import.meta.url)) {
  main();
}